File: javascript/cl/JSAN/src/ast/declaration/functionDeclaration.js
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2018 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

import * as globals from '../../globals.js';

const factory = globals.getFactory();

export default function (node, parent, firstVisit) {
    if (firstVisit) {
        if (globals.getWrapperOfNode(node) !== undefined) {
            return;
        }
        const functionDeclaration = factory.createFunctionDeclarationWrapper();
        globals.setPositionInfo(node, functionDeclaration);
        functionDeclaration.setGenerator(node.generator);
        functionDeclaration.setAsync(node.async);
        return functionDeclaration;
    } else {
        const functionDeclarationWrapper = globals.getWrapperOfNode(node);

        if (node.params != null) {
            for (let i = 0; i < node.params.length; i++) {
                if (node.params[i] != null) {
                    globals.safeSet(functionDeclarationWrapper, "addParams", node.params[i], "FUNCTIONDECLARATION - Could not add param");
                }
            }
        }

        if (node.declaration != null) {
            globals.safeSet(functionDeclarationWrapper, "setDeclaration", node.declaration, "FUNCTIONDECLARATION - Could not set declaration!");
        }

        if (node.body != null) {
            globals.safeSet(functionDeclarationWrapper, "setBody", node.body, "FUNCTIONDECLARATION - Could not set body!");
        }

        if (node.id != null) {
            globals.safeSet(functionDeclarationWrapper, "setIdentifier", node.id, "FUNCTIONDECLARATION - Could not set identifier!");
        }

    }
}

File: lib/javascript/src/declaration/FunctionDeclaration.cpp
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2018 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

#include "javascript/inc/javascript.h"
#include "javascript/inc/Common.h"
#include "common/inc/WriteMessage.h"

#include "javascript/inc/messages.h"
#include <algorithm>
#include <string.h>
#include "common/inc/math/common.h"


namespace columbus { namespace javascript { namespace asg {

typedef boost::crc_32_type  Crc_type;

namespace declaration { 
  FunctionDeclaration::FunctionDeclaration(NodeId _id, Factory *_factory) :
         Positioned(_id, _factory),
    Function(_id, _factory),
    Declaration(_id, _factory)
  {
  }

  FunctionDeclaration::~FunctionDeclaration() {
  }

  void FunctionDeclaration::prepareDelete(bool tryOnVirtualParent){
    if (tryOnVirtualParent) {
      base::Positioned::prepareDelete(false);
    }
    statement::Function::prepareDelete(false);
    declaration::Declaration::prepareDelete(false);
  }

  NodeKind FunctionDeclaration::getNodeKind() const {
    return ndkFunctionDeclaration;
  }

  bool FunctionDeclaration::setEdge(EdgeKind edgeKind, NodeId edgeEnd, bool tryOnVirtualParent) {
    if (tryOnVirtualParent) {
      if (base::Positioned::setEdge(edgeKind, edgeEnd, false)) {
        return true;
      }
    }
    if (statement::Function::setEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    if (declaration::Declaration::setEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    return false;
  }

  bool FunctionDeclaration::removeEdge(EdgeKind edgeKind, NodeId edgeEnd, bool tryOnVirtualParent) {
    if (tryOnVirtualParent) {
      if (base::Positioned::removeEdge(edgeKind, edgeEnd, false)) {
        return true;
      }
    }
    if (statement::Function::removeEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    if (declaration::Declaration::removeEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    return false;
  }

  void FunctionDeclaration::accept(Visitor &visitor) const {
    visitor.visit(*this);
  }

  void FunctionDeclaration::acceptEnd(Visitor &visitor) const {
    visitor.visitEnd(*this);
  }

  double FunctionDeclaration::getSimilarity(const base::Base& base){
    if(base.getNodeKind() == getNodeKind()) {
      const FunctionDeclaration& node = dynamic_cast<const FunctionDeclaration&>(base);
      double matchAttrs = 0;
      if(node.getGenerator() == getGenerator()) ++matchAttrs;
      if(node.getAsync() == getAsync()) ++matchAttrs;
      return matchAttrs / (2 / (1 - Common::SimilarityMinimum)) + Common::SimilarityMinimum;
    } else {
      return 0.0;
    }
  }

  void FunctionDeclaration::swapStringTable(RefDistributorStrTable& newStrTable, std::map<Key,Key>& oldAndNewStrKeyMap ){
    std::map<Key,Key>::iterator foundKeyId;

  }

  NodeHashType FunctionDeclaration::getHash(std::set<NodeId>& travNodes) const {
    if (hashOk) return nodeHashCache;
    common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_BEGIN,this->getId()); 
    if (travNodes.count(getId())>0) {
      common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_SKIP);
      return 0;
    }
    travNodes.insert(getId());
    Crc_type resultHash;
    resultHash.process_bytes( "declaration::FunctionDeclaration", strlen("declaration::FunctionDeclaration"));
    common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_END,resultHash.checksum()); 
    nodeHashCache = resultHash.checksum();
    hashOk = true;
    return nodeHashCache;
  }

  void FunctionDeclaration::save(io::BinaryIO &binIo,bool withVirtualBase  /*= true*/) const {
    if (withVirtualBase)
      Positioned::save(binIo,false);

    Function::save(binIo,false);

    Declaration::save(binIo,false);

  }

  void FunctionDeclaration::load(io::BinaryIO &binIo, bool withVirtualBase /*= true*/) {
    if (withVirtualBase)
      Positioned::load(binIo, false);

    Function::load(binIo,false);

    Declaration::load(binIo,false);

  }


}


}}}
